backend/wmg/api/rollup.py

```diff
@@ -0,0 +1,130 @@
+import owlready2
+import pandas as pd
+import numpy as np
+import numba as nb
+from functools import lru_cache
+from backend.wmg.data.constants import CL_BASIC_PERMANENT_URL_OWL
+
+
+# ontology object
+onto = None
+
+
+def find_descendants_per_cell_type(cell_types):
+    """
+    Find descendants for each cell type in the input list.
+
+    Parameters
+    ----------
+    cell_types : list
+        List of cell types (cell type ontology term IDs) to find descendants for
+
+    Returns
+    -------
+    descendants_per_cell_type : list
+        List of lists of descendants for each cell type in the input list.
+    """
+
+    global onto
+    if not onto:
+        onto = owlready2.get_ontology(CL_BASIC_PERMANENT_URL_OWL)
+        onto.load()
+
+    # cache finding descendants per cell type
+    @lru_cache(maxsize=None)
+    def _descendants(cell_type):
+        cell_type_iri = cell_type.replace(":", "_")
+        entity = onto.search_one(iri=f"http://purl.obolibrary.org/obo/{cell_type_iri}")
```

Comment:
I think you could save the ontology file locally and do the lookups on a local file, and that might increase performance.